gctree/isotype.py

```diff
@@ -0,0 +1,432 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import re
+import random
+import ete3
+from gctree.utils import hamming_distance
+import pickle
+import argparse
+from pathlib import Path
+import warnings
+from typing import Dict, List
+
+
+class IsotypeTemplate:
+    def __init__(self, isotype_order):
+        self.order = isotype_order
+
+    def new(self, isotype_name):
+        return Isotype(self.order, isotype_name)
+
+
+class Isotype:
+    # From https://doi.org/10.7554/eLife.16578.012, for humans:
+    # order = ["M", "G3", "A1", "G2", "G4", "E", "A2"]
+
+    def __init__(self, order: List[str], isotype_name: str):
+        self.order = order
+        if isotype_name == "?":
+            self.isotype = None
+        else:
+            try:
+                self.isotype = self.order.index(isotype_name)
+            except ValueError as e:
+                raise ValueError(
+                    "Unrecognized isotype name. Check that default"
+                    " or provided isotype name list contains all"
+                    " observed isotypes\n" + str(e)
+                )
+
+    def isbefore(self, t: "Isotype") -> bool:
+        return self.isotype <= t.isotype
```

Comment:
A possible runtime error here will occur (and possibly silently) if `self.order` is not identical to `t.order`. Maybe we should check for this?